### PR TITLE
Add a check to verify secret key change in database.

### PR DESCRIPTION
**Pull request by Wathsara**

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 1]
```diff
@@ -395,6 +396,33 @@ private void checkTotpEnabled(AuthenticationContext context, String username) th
         }
     }
 
+    /**
+     * Update the secret key in the context with value in database before validate.
+     *
+     * @param context  Authenticated context.
+     * @param username Authenticated users' username
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:403]: Missing a `.` at the end of the statement.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 2]
```diff
@@ -395,6 +396,33 @@ private void checkTotpEnabled(AuthenticationContext context, String username) th
...
+
+        try {
+            String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+            UserRealm userRealm = TOTPUtil.getUserRealm(username);
+            if (userRealm != null) {
+                Map<String, String> userClaimValues = userRealm.getUserStoreManager()
+                        .getUserClaimValues(tenantAwareUsername,
+                                new String[]{TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL}, null);
+                String secretKey = userClaimValues.get(TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL);
+                if (StringUtils.isNotEmpty(secretKey)) {
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:417]: Do we need to check for a blank here?

Comment (Wathsara) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:417]: Yes. This flow is only for the users that have no secret key available in database when starting the authentication flow. Therefore we keep the generated secret key in the context, only if database get change from some other process we override the context info.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java (hunk 1 above)
Comment (Wathsara) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java]: Address with - https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/113/commits/e27180b645c234705cdc784dd90e0eaef49943fc

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 3]
```diff
@@ -395,6 +396,33 @@ private void checkTotpEnabled(AuthenticationContext context, String username) th
...
+                    context.setProperty(TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL, secretKey);
+                }
+            }
+        } catch (UserStoreException e) {
+            throw new AuthenticationFailedException("Error while getting TOTP secret key : " + username, e);
```
Comment (mevan-karu) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:422]: Why are setting username as the secret key?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 4]
```diff
@@ -375,6 +375,7 @@ private void checkTotpEnabled(AuthenticationContext context, String username) th
 
         if (context.getProperty(TOTPAuthenticatorConstants.ENABLE_TOTP) != null && Boolean
                 .valueOf(context.getProperty(TOTPAuthenticatorConstants.ENABLE_TOTP).toString())) {
+            checkForUpdatedSecretKey(context,username);
```
Comment (mevan-karu) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:378]: Formatting

**Review by mevan-karu — APPROVED** [review]

**Review by IsuraD — APPROVED** [review]

**Pull request merged**